packages/serverpod_test/lib/src/serverpod_test.dart

```diff
@@ -0,0 +1,415 @@
+          await testSession.destroy();
+        }
+        allTestSessions.clear();
+
+        await testServerpod.shutdown();
+      });
+
+      testClosure(testServerpod.testEndpoints, mainTestSession);
+    });
+  };
+}
+
+class _TransactionManager {
+  final List<String> _savePointIds = [];
+
+  Transaction? _transaction;
+
+  late Completer _endTransactionScopeCompleter;
+
+  late Session serverpodSession;
+
+  _TransactionManager(this.serverpodSession);
+
+  Future<Transaction> createTransaction() async {
+    if (_transaction != null) {
+      throw StateError('Transaction already exists.');
+    }
+
+    _endTransactionScopeCompleter = Completer();
+    var transactionStartedCompleter = Completer();
+    late Transaction localTransaction;
+
+    unawaited(
+      serverpodSession.db.transaction((newTransaction) async {
+        localTransaction = newTransaction;
+
+        transactionStartedCompleter.complete();
+
+        await _endTransactionScopeCompleter.future
+            .timeout(const Duration(seconds: 60), onTimeout: () async {
+          await newTransaction.cancel();
+        });
+      }),
+    );
+
+    await transactionStartedCompleter.future;
+
+    _transaction = localTransaction;
+
+    return localTransaction;
+  }
+
+  Future<void> cancelTransaction() async {
+    var localTransaction = _transaction;
+    if (localTransaction == null) {
+      throw StateError('No ongoing transaction.');
+    }
+
+    await localTransaction.cancel();
+    _endTransactionScopeCompleter.complete();
+    _transaction = null;
+  }
+
+  Future<void> pushSavePoint() async {
+    if (_transaction == null) {
+      throw StateError('No ongoing transaction.');
+    }
+
+    var savePointId = _getNextSavePointId();
+    _savePointIds.add(savePointId);
+
+    await serverpodSession.db.unsafeExecute(
+      'SAVEPOINT $savePointId;',
+      transaction: _transaction,
+    );
+  }
+
+  String _getNextSavePointId() {
+    var postgresCompatibleRandomString =
+        const Uuid().v4obj().toString().replaceAll(RegExp(r'-'), '_');
+    var savePointId = 'savepoint_$postgresCompatibleRandomString';
+
+    return savePointId;
+  }
+
+  Future<void> popSavePoint() async {
+    if (_transaction == null) {
+      throw StateError('No ongoing transaction.');
+    }
+
+    if (_savePointIds.isEmpty) {
+      throw StateError('No previous savepoint to rollback to.');
+    }
+
+    await serverpodSession.db.unsafeExecute(
+      'ROLLBACK TO SAVEPOINT ${_savePointIds.removeLast()};',
+      transaction: _transaction,
+    );
+  }
+}
+
+/// Test tools helper to not leak exceptions from awaitable functions.
+/// Used by the generated code.
+Future<T> callAwaitableFunctionAndHandleExceptions<T>(
+  Future<T> Function() call,
+) async {
+  try {
+    return await call();
+  } catch (e) {
+    throw _getException(e);
+  }
+}
+
+/// Test tools helper to not leak exceptions from functions that return streams.
+/// Used by the generated code.
```

Comment:
Suggestion: Add information that this enables the streams execution to start without having to wait for a listener.

One benefit in this case was to enable streams to run without having to think about subscriptions. We should clarify that here for future devs.

By the way, do we have a test validating this specific case?